src/controller_examples/zookeeper_controller/spec/zookeeper_api.rs

```diff
@@ -8,92 +8,205 @@ use vstd::{prelude::*, string::*};
+// Note that it uses the uid to avoid querying the data belonging to the old stateful set object.
+pub open spec fn handle_exists(
+    name: StringView, namespace: StringView, path: Seq<StringView>, resources: StoredState, state: ZKState
+) -> (ZKState, ZKAPIExistsResultView) {
+    let key = ObjectRef { kind: Kind::StatefulSetKind, namespace: namespace, name: name };
+    if !validate(name, namespace, path, resources) {
+        (state, ZKAPIExistsResultView{res: Err(Error::ZKNodeExistsFailed)})
+    } else {
+        let addr = ZKNodeAddr::new(name, namespace, resources[key].metadata.uid.get_Some_0(), path);
+        if !state.data.dom().contains(addr) {
+            (state, ZKAPIExistsResultView{res: Ok(None)})
+        } else {
+            let version = state.data[addr].1;
+            (state, ZKAPIExistsResultView{res: Ok(Some(version))})
+        }
+    }
+}
+
+// handle_create models the behavior of the zookeeper server handling the create request.
+// The creation succeeds only when (1) the node does not exist yet and (2) the parent node exists.
+pub open spec fn handle_create(
+    name: StringView, namespace: StringView, path: Seq<StringView>, data: ZKNodeValue, resources: StoredState, state: ZKState
+) -> (ZKState, ZKAPICreateResultView) {
+    let key = ObjectRef { kind: Kind::StatefulSetKind, namespace: namespace, name: name };
+    if !validate(name, namespace, path, resources) {
+        (state, ZKAPICreateResultView{res: Err(Error::ZKNodeCreateFailed)})
+    } else {
+        let addr = ZKNodeAddr::new(name, namespace, resources[key].metadata.uid.get_Some_0(), path);
+        if !state.data.dom().contains(addr) {
+            if path.len() > 1 && !state.data.dom().contains(addr.parent_addr()) {
+                (state, ZKAPICreateResultView{res: Err(Error::ZKNodeCreateFailed)})
+            } else {
+                let state_prime = ZKState {
+                    data: state.data.insert(addr, (data, 0)),
+                };
+                (state_prime, ZKAPICreateResultView{res: Ok(())})
+            }
+        } else {
+            (state, ZKAPICreateResultView{res: Err(Error::ZKNodeCreateAlreadyExists)})
+        }
+    }
+}
+
+// handle_set_data models the behavior of the zookeeper server handling the set data request.
+// To set the data, the node needs to exist and the provided version number must match the current version of the node.
+pub open spec fn handle_set_data(
+    name: StringView, namespace: StringView, path: Seq<StringView>, data: ZKNodeValue, version: ZKNodeVersion, resources: StoredState, state: ZKState
+) -> (ZKState, ZKAPISetDataResultView) {
+    let key = ObjectRef { kind: Kind::StatefulSetKind, namespace: namespace, name: name };
+    if !validate(name, namespace, path, resources) {
+        (state, ZKAPISetDataResultView{res: Err(Error::ZKNodeSetDataFailed)})
+    } else {
+        let addr = ZKNodeAddr::new(name, namespace, resources[key].metadata.uid.get_Some_0(), path);
```

Comment:
Sorry, I mean in this case, the node should already exist?

Reply:
Yes: https://github.com/bonifaido/rust-zookeeper/blob/master/src/zookeeper.rs#L472-L486

Reply:
So why you still check the existence here?

Reply:
If we don't check, how do we know whether the node exists? Just to clarify this spec function is used to model the behavior of how zookeeper server handles the set_data request.

Reply:
In other words, this spec function should precisely specify this https://github.com/bonifaido/rust-zookeeper/blob/master/src/zookeeper.rs#L472-L486

Reply:
Because when you call this function, the node must exist.

Reply:
By "you", do you mean any possible user, or just our zk controller?